crates/chia-datalayer/src/lib.rs

```diff
@@ -0,0 +1,1259 @@
+#[cfg(feature = "py-bindings")]
+use pyo3::{buffer::PyBuffer, pyclass, pymethods, PyResult};
+
+use clvmr::sha2::Sha256;
+use std::cmp::Ordering;
+use std::collections::{HashMap, VecDeque};
+use std::iter::IntoIterator;
+use std::ops::Range;
+
+// TODO: clearly shouldn't be hard coded
+const METADATA_SIZE: usize = 2;
+// TODO: clearly shouldn't be hard coded
+const DATA_SIZE: usize = 44;
+const BLOCK_SIZE: usize = METADATA_SIZE + DATA_SIZE;
+
+type TreeIndex = u32;
+type Parent = Option<TreeIndex>;
+type Hash = [u8; 32];
+type BlockBytes = [u8; BLOCK_SIZE];
+type KvId = u64;
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+#[repr(u8)]
+pub enum NodeType {
+    Internal = 0,
+    Leaf = 1,
+}
+
+impl NodeType {
+    pub fn from_u8(value: u8) -> Result<Self, String> {
+        // TODO: identify some useful structured serialization tooling we use
+        // TODO: find a better way to tie serialization values to enumerators
+        match value {
+            // ha!  feel free to laugh at this
+            x if (NodeType::Internal as u8 == x) => Ok(NodeType::Internal),
+            x if (NodeType::Leaf as u8 == x) => Ok(NodeType::Leaf),
+            other => panic!("unknown NodeType value: {other}"),
+        }
+    }
+
+    pub fn to_u8(&self) -> u8 {
+        match self {
+            NodeType::Internal => NodeType::Internal as u8,
+            NodeType::Leaf => NodeType::Leaf as u8,
+        }
+    }
+}
+
+// impl NodeType {
+//     const TYPE_TO_VALUE: HashMap<NodeType, u8> = HashMap::from([
+//         (NodeType::Internal, 0),
+//         (NodeType::Leaf, 1),
+//     ]);
+//
+//     fn value(&self) -> u8 {
+//         let map = Self::TYPE_TO_VALUE;
+//         // TODO: this seems pretty clearly the wrong way, probably
+//         let value = map.get(self);
+//         if value.is_some() {
+//             return 3;
+//         }
+//         panic!("no value for NodeType: {self:?}");
+//     }
+// }
+
+fn internal_hash(left_hash: Hash, right_hash: Hash) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(b"\x02");
+    hasher.update(left_hash);
+    hasher.update(right_hash);
+
+    hasher.finalize()
+}
+
+pub struct DotLines {
+    nodes: Vec<String>,
+    connections: Vec<String>,
+    pair_boxes: Vec<String>,
+}
+
+impl Default for DotLines {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DotLines {
+    pub fn new() -> Self {
+        Self {
+            nodes: vec![],
+            connections: vec![],
+            pair_boxes: vec![],
+        }
+    }
+
+    pub fn push(&mut self, mut other: DotLines) {
+        self.nodes.append(&mut other.nodes);
+        self.connections.append(&mut other.connections);
+        self.pair_boxes.append(&mut other.pair_boxes);
+    }
+
+    pub fn dump(&mut self) -> String {
+        // TODO: consuming itself, secretly
```

Comment:
that's a todo saying i don't like how this works.  but i didn't get it working how i wanted and this worked sufficiently and i moved on to other bits with a note to revisit.  the appends are consuming the insides of `self` out of it right now.  so `self` is now garbage, but rust doesn't know that because of the function signature.  but i don't actually want to consume it in any form.

Reply:
if you just want to consume the vectors of `self`, you could use `self.nodes.drain()`.

Reply:
Sounds like he's wanting to clone, rather than drain?